geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeatureStore.scala

```diff
@@ -0,0 +1,133 @@
+    val sew = startWeek.getWeeks
+    val endWeek = CassandraPrimaryKey.epochWeeks(interval.getEnd)
+    val eew = endWeek.getWeeks
+
+    val rows = (sew to eew).map { dt => getRowKeys(lx, ly, ux, uy, interval, sew, eew, dt) }
+
+    val features = contentState.builderPool.withResource { builder =>
+      val futures = rows.flatMap { case ((s, e), rowRanges) => executeQueryForContiguousRowRange(s, e, rowRanges) }
+      futures.flatMap { case (r, contains, fut) =>
+        postProcessResults(query, builder, contains, fut)
+      }
+    }
+
+    features.iterator
+  }
+
+  def postProcessResults(query: Query, builder: SimpleFeatureBuilder, contains: Boolean, fut: ResultSetFuture): List[SimpleFeature] = {
+    val featureIterator = fut.get().iterator().map { r => convertRowToSF(r, builder) }
+    val filt = query.getFilter
+    val iter =
+      if (!contains) featureIterator.filter(f => filt.evaluate(f)).toList
+      else featureIterator.toList
+    iter
+  }
+
+  def executeQueryForContiguousRowRange(s: Int, e: Int, rowRanges: Seq[Int]): Seq[(Int, Boolean, ResultSetFuture)] = {
+    rowRanges.flatMap { r =>
+      val CassandraPrimaryKey.Key(_, _, _, _, z) = CassandraPrimaryKey.unapply(r)
+      val (minx, miny, maxx, maxy) = CassandraPrimaryKey.SFC2D.bound(z)
+      val z3ranges =
+        org.locationtech.geomesa.cassandra.data.CassandraPrimaryKey.SFC3D.ranges((minx, maxx), (miny, maxy), (s, e))
+
+      z3ranges.map { case (l, u, contains) =>
+        val q = contentState.GEO_TIME_QUERY.bind(r: Integer, l: java.lang.Long, u: java.lang.Long)
+        (r, contains, contentState.session.executeAsync(q))
+      }
+    }
+  }
+
+  def getRowKeys(lx: Double, ly: Double, ux: Double, uy: Double, interval: Interval, sew: Int, eew: Int, dt: Int): ((Int, Int), Seq[Int]) = {
+    val dtshift = dt << 16
+    val dtg = new DateTime(0).plusWeeks(dt)
+
+    val seconds =
+      if (dt != sew && dt != eew) {
+        (0, CassandraPrimaryKey.ONE_WEEK_IN_SECONDS)
+      } else {
+        val starts =
+          if (dt == sew) CassandraPrimaryKey.secondsInCurrentWeek(interval.getStart)
+          else 0
+        val ends =
+          if (dt == eew) CassandraPrimaryKey.secondsInCurrentWeek(interval.getEnd)
+          else CassandraPrimaryKey.ONE_WEEK_IN_SECONDS
+        (starts, ends)
+      }
+    val zranges = org.locationtech.geomesa.cassandra.data.CassandraPrimaryKey.SFC2D.toRanges(lx, ly, ux, uy)
+    val shiftedRanges = zranges.flatMap { case (l, u, _) => (l to u).map { i => (dtshift + i).toInt } }
+    (seconds, shiftedRanges)
+  }
+
+  def getAllFeatures = {
+    contentState.builderPool.withResource { builder =>
+      contentState.session.execute(contentState.ALL_QUERY.bind()).iterator().map { r => convertRowToSF(r, builder) }
+    }
+  }
+
+  def convertRowToSF(r: Row, builder: SimpleFeatureBuilder): SimpleFeature = {
+    val attrs = contentState.deserializers.zipWithIndex.map { case (d, idx) => d.deserialize(r.getObject(idx + 1)) }
```

Comment:
zip with index is semi-expensive, might be good to pull it out and not do it for every feature.